### PR TITLE
Add simple tests for entry/exit actions on source & proper descendant…

### PR DESCRIPTION
**Pull request by Andarist**
Related to: https://github.com/davidkpiano/xstate/issues/700 , just adding some tests for this behavior in the core. We've already enabled [test505](https://www.w3.org/Voice/2013/scxml-irp/505/test505.txml) that is testing this part of the spec, although it only tests `exit` handlers (which I guess is more important as you can't reentry proper descendants without exiting them first). That test is somewhat complicated (like most of the tests in that test suite) so I've decided not to use its direct translation which is this:
```js
Machine({
  context: {
    Var1: 0,
    Var2: 0,
    Var3: 0,
  },
  initial: "s1",
  states: {
    s1: {
      initial: "s11",
      onEntry: [raise("foo"), raise("bar")],
      onExit: assign({
        Var1: (ctx) => ctx.Var1 + 1,
      }),
      states: {
        s11: {
          onExit: assign({
            Var2: (ctx) => ctx.Var2 + 1,
          }),
        },
      },
      on: {
        foo: {
          target: ".s11",
          actions: assign({
            Var3: (ctx) => ctx.Var3 + 1,
          }),
          internal: true,
        },
        bar: [
          {
            cond: (ctx) => ctx.Var3 === 1,
            target: "s2",
          },
          {
            target: "#fail",
          },
        ],
      },
    },
    s2: {
      always: [
        {
          cond: (ctx) => ctx.Var1 === 1,
          target: "s3",
        },
        {
          target: "#fail",
        },
      ],
    },
    s3: {
      always: [
        {
          cond: (ctx) => ctx.Var2 === 2,
          target: "#pass",
        },
        {
          target: "#fail",
        },
      ],
    },
    pass: {
      id: "pass",
      type: "final",
    },
    fail: {
      id: "fail",
      type: "final",
    },
  },
});
```

### CODE REVIEW

File: packages/core/test/internalTransitions.test.ts [hunk 1]
```diff
@@ -140,4 +140,106 @@ describe('internal transitions', () => {
...
+        sourceStateEntries: 0,
+        directDescendantEntries: 0,
+        deepDescendantEntries: 0
+      },
+      initial: 'a1',
+      states: {
+        a1: {
+          initial: 'a11',
+          entry: assign({
+            sourceStateEntries: (ctx) => ++ctx.sourceStateEntries
```
Comment (davidkpiano) [inline packages/core/test/internalTransitions.test.ts:160]:
```suggestion
            sourceStateEntries: (ctx) => ctx.sourceStateEntries + 1
```

File: packages/core/test/internalTransitions.test.ts [hunk 2]
```diff
@@ -140,4 +140,106 @@ describe('internal transitions', () => {
...
+          }),
+          states: {
+            a11: {
+              initial: 'a111',
+              entry: assign({
+                directDescendantEntries: (ctx) => ++ctx.directDescendantEntries
```
Comment (davidkpiano) [inline packages/core/test/internalTransitions.test.ts:166]:
```suggestion
                directDescendantEntries: (ctx) => ctx.directDescendantEntries + 1
```

File: packages/core/test/internalTransitions.test.ts [hunk 3]
```diff
@@ -140,4 +140,106 @@ describe('internal transitions', () => {
...
+              }),
+              states: {
+                a111: {
+                  entry: assign({
+                    deepDescendantEntries: (ctx) => ++ctx.deepDescendantEntries
```
Comment (davidkpiano) [inline packages/core/test/internalTransitions.test.ts:171]:
```suggestion
                    deepDescendantEntries: (ctx) => ctx.deepDescendantEntries + 1
```

File: packages/core/test/internalTransitions.test.ts [hunk 4]
```diff
@@ -140,4 +140,106 @@ describe('internal transitions', () => {
...
+        sourceStateExits: 0,
+        directDescendantExits: 0,
+        deepDescendantExits: 0
+      },
+      initial: 'a1',
+      states: {
+        a1: {
+          initial: 'a11',
+          exit: assign({
+            sourceStateExits: (ctx) => ++ctx.sourceStateExits
```
Comment (davidkpiano) [inline packages/core/test/internalTransitions.test.ts:211]:
```suggestion
            sourceStateExits: (ctx) => ctx.sourceStateExits + 1
```

File: packages/core/test/internalTransitions.test.ts [hunk 5]
```diff
@@ -140,4 +140,106 @@ describe('internal transitions', () => {
...
+          }),
+          states: {
+            a11: {
+              initial: 'a111',
+              exit: assign({
+                directDescendantExits: (ctx) => ++ctx.directDescendantExits
```
Comment (davidkpiano) [inline packages/core/test/internalTransitions.test.ts:217]:
```suggestion
                directDescendantExits: (ctx) => ctx.directDescendantExits + 1
```

File: packages/core/test/internalTransitions.test.ts [hunk 6]
```diff
@@ -140,4 +140,106 @@ describe('internal transitions', () => {
...
+              }),
+              states: {
+                a111: {
+                  exit: assign({
+                    deepDescendantExits: (ctx) => ++ctx.deepDescendantExits
```
Comment (davidkpiano) [inline packages/core/test/internalTransitions.test.ts:222]:
```suggestion
                    deepDescendantExits: (ctx) => ctx.deepDescendantExits + 1
```

**Comment by Andarist** [comment]
@davidkpiano amended with requested changes

**Pull request merged**